gc/base/standard/ConcurrentGCSATB.cpp

```diff
@@ -1,145 +1,405 @@
+{
+	/* ..and then tearDown our super class */
+	MM_ConcurrentGC::tearDown(env);
+}
+
+void
+MM_ConcurrentGCSATB::reportConcurrentHalted(MM_EnvironmentBase *env)
+{
+	OMRPORT_ACCESS_FROM_ENVIRONMENT(env);
+
+	/* Redo Trace points for SATB */
+
+	TRIGGER_J9HOOK_MM_PRIVATE_CONCURRENT_HALTED(
+		_extensions->privateHookInterface,
+		env->getOmrVMThread(),
+		omrtime_hires_clock(),
+		J9HOOK_MM_PRIVATE_CONCURRENT_HALTED,
+		(uintptr_t)_stats.getExecutionModeAtGC(),
+		_stats.getTraceSizeTarget(),
+		_stats.getTotalTraced(),
+		_stats.getMutatorsTraced(),
+		_stats.getConHelperTraced(),
+		UDATA_MAX,
+		UDATA_MAX,
+		_stats.getConcurrentWorkStackOverflowOcurred(),
+		_stats.getConcurrentWorkStackOverflowCount(),
+		UDATA_MAX,
+		_concurrentDelegate.reportConcurrentScanningMode(env),
+		(uintptr_t)_markingScheme->getWorkPackets()->tracingExhausted()
+	);
+}
+
+uintptr_t
+MM_ConcurrentGCSATB::localMark(MM_EnvironmentBase *env, uintptr_t sizeToTrace)
+{
+	omrobjectptr_t objectPtr;
+	uintptr_t gcCount = _extensions->globalGCStats.gcCount;
+
+	env->_workStack.reset(env, _markingScheme->getWorkPackets());
+	Assert_MM_true(NULL == env->_cycleState);
+	Assert_MM_true(CONCURRENT_OFF < _stats.getExecutionMode());
+	Assert_MM_true(_concurrentCycleState._referenceObjectOptions == MM_CycleState::references_default);
+	env->_cycleState = &_concurrentCycleState;
+
+	uintptr_t sizeTraced = 0;
+	while (NULL != (objectPtr = (omrobjectptr_t)env->_workStack.popNoWait(env))) {
+		/* Check for array scanPtr..if we find one ignore it*/
+		if (((uintptr_t)objectPtr) & PACKET_ARRAY_SPLIT_TAG) {
+			continue;
+		} else {
+			/* Else trace the object */
+			sizeTraced += _markingScheme->scanObject(env, objectPtr, SCAN_REASON_PACKET, (sizeToTrace - sizeTraced));
+		}
+
+		/* Have we done enough tracing ? */
+		if (sizeTraced >= sizeToTrace) {
+			break;
+		}
+
+		/* Before we do any more tracing check to see if GC is waiting */
+		if (env->isExclusiveAccessRequestWaiting()) {
+			/* suspend con helper thread for pending GC */
+			uintptr_t conHelperRequest = switchConHelperRequest(CONCURRENT_HELPER_MARK, CONCURRENT_HELPER_WAIT);
+			Assert_MM_true(CONCURRENT_HELPER_MARK != conHelperRequest);
+			break;
+		}
+	}
+
+	/* Pop the top of the work packet if its a partially processed array tag */
+	if (((uintptr_t)((omrobjectptr_t)env->_workStack.peek(env))) & PACKET_ARRAY_SPLIT_TAG) {
+		env->_workStack.popNoWait(env);
+	}
+
+	/* STW collection should not occur while localMark is working */
+	Assert_MM_true(gcCount == _extensions->globalGCStats.gcCount);
+
+	flushLocalBuffers(env);
+	env->_cycleState = NULL;
+
+	return sizeTraced;
+}
+
+#if defined(OMR_GC_MODRON_SCAVENGER)
+/**
+ * Process event from an external GC (Scavenger) when old-to-old reference is created.
+ * @param objectPtr  Parent old object that has a reference to a child old object
+ */
+void
+MM_ConcurrentGCSATB::oldToOldReferenceCreated(MM_EnvironmentBase *env, omrobjectptr_t objectPtr)
+{}
```

Comment:
I believe curly brackets should be broken into 2 lines or you can inline in *.hpp, and then keep everything including declaration in one line